Port cartesian_impedance_controller to franka_ros2 example controllers

sp-sophia-labs opened the pull request:
**Description:**  

This PR offers a high level port of Cartesian impedance example controller provided in  [Franka ROS1](https://github.com/frankaemika/franka_ros) and requests feedback. 

<p align="center">
  <img src="https://github.com/frankaemika/franka_ros2/assets/111747728/64e604b9-b5e4-440d-9efb-3ed6ac87f9ce">
</p>
<p align="center" style="margin-top: -10px; font-size: 4px;">
    <em>ros humble run test</em>
</p>

The port was basically based on : 
- ROS2 Franka example controllers for global architecture of controller and getting robot information (Robot State and Robot Model). 
- ROS1 Cartesian example controller example for the theoretical part (basically real-time update() method), all computation is pretty much ported with no changes. 

**Architecture:**

Controller class is implemented based on ROS2 Control Controller Interface Class : 
- Member variables initialized in the constructor header.
- init() method calls ControllerInterface::init() to initialize the lifecycle of the controller. Following this, a declaration of 
all parameters defining their default values are.
- Get Robot State and Robot Model Interfaces in State_interface_configuration() : 

`

    const std::string k_robot_state_interface_name{"robot_state"};
    const std::string k_robot_model_interface_name{"robot_model"};

    for (const auto& franka_robot_model_name : franka_robot_model_->get_state_interface_names()) {
          state_interfaces_config.names.push_back(franka_robot_model_name);
     }

    for (const auto& franka_robot_state_name : franka_robot_state_->get_state_interface_names()) {
          state_interfaces_config.names.push_back(franka_robot_state_name);
     }

`

- Get Effort Command interface in Command_interface_configuration() :

`

    for (int i = 1; i <= num_joints; i++) 
    {
          command_interfaces_config.names.push_back(arm_id_ + "_joint" + std::to_string(i) + "/effort");     
     }

`

- Real-time update() method is pretty much ported from ROS1 version (requires release build : DCMAKE_BUILD_TYPE=Release) 

- lifecycle management methods : 
1. on_configure() and on_deactivate() : Nothing special here, first to configures controller and second is called when controller is deactivated (stopped).

2. on_activate() - called when controller is activated (started), I think this method is the equivalent functionality to the starting() method in ROS 1, since it is called when a controller is activated, and this is a suitable place to perform actions 
that need to occur just before the controller starts updating its state.


**Notes :** 
1. For Equilibrium pose, I first set it as initial state, then published through a separated node 
2. Compliance Parameters are set and updated through the code, Not dynamically configured yet. 
3. Controller Have been tested on Franka Emika research3 FR3.   

**Questions :** 

I need feedback/confirmation about those points if possible : 
1. Getting robot state
2. Getting Initial robot state and Setting parameters on Activate() method 


Thanks in advance !

zakxbt committed:
import cartesian_impedance_controller to franka_ros2 example controllers

zakxbt committed:
clean up headers

zakxbt committed:
saturation

sp-sophia-labs commented:
Friendly ping @BarisYazici

peterdavidfagan commented:
@sp-sophia-labs has this controller been tested on the hardware platform? Also interested in this effort.

BarisYazici commented:
Sorry for the delay! I will try to review in short the upcoming days

BarisYazici commented on franka_example_controllers/src/cartesian_impedance_example_controller.cpp:
```diff
@@ -0,0 +1,243 @@
+    Eigen::MatrixXd jacobian_transpose_pinv;
+
+    try {
+      // Release Build : colcon build --cmake-args -DCMAKE_BUILD_TYPE=Release
```
some leftover I guess

BarisYazici commented:
Can you also create the bringup launch file that starts this controller?

BarisYazici commented on franka_example_controllers/src/cartesian_impedance_example_controller.cpp:
```diff
@@ -0,0 +1,243 @@
+#include "franka_example_controllers/cartesian_impedance_example_controller.hpp"
+#include "franka_example_controllers/pseudo_inverse.hpp"
```
Header file is called pseudo_inversion.hpp

BarisYazici commented on franka_example_controllers/src/cartesian_impedance_example_controller.cpp:
```diff
@@ -0,0 +1,243 @@
+    // Equilibrium pose subscription
+    sub_equilibrium_pose_ = get_node()->create_subscription<geometry_msgs::msg::PoseStamped>(
+      "equilibrium_pose", 20, 
+      std::bind(&FSMImpedanceController::equilibriumPoseCallback, this, std::placeholders::_1));
```
where is this &FSMImpedanceController coming from? Could you also provide the code for that?

sp-sophia-labs replied:
This was related to local class on our project, now equilibriumPoseCallback is memeber of &CartesianImpedanceExampleController

zakxbt committed:
Cartesian Impedance Example Controller config

zakxbt committed:
clenup leftover & fix equilibriumPoseCallback call & pseudo inversion header

zakxbt committed:
Cartesian Impedance Example launch file

sp-sophia-labs commented:
@BarisYazici Thank you for your feedback! I have made the requested changes:
- Updated equilibriumPoseCallback call.   
- Cleaned up leftover and header call.
- Added launch for test.

Please let me know if there are any further adjustments needed or if there's anything else you'd like me to address

AndreasKuhner commented on franka_example_controllers/include/franka_example_controllers/cartesian_impedance_example_controller.hpp:
```diff
@@ -0,0 +1,92 @@
+#include "franka_example_controllers/visibility_control.h"
+
+
+#define IDENTITY Eigen::MatrixXd::Identity(6,6)
```
A `#define` looks a bit harsh. Maybe you could just define a constant for it - best case only in the .cpp if you need it there.

AndreasKuhner commented on franka_example_controllers/include/franka_example_controllers/cartesian_impedance_example_controller.hpp:
```diff
@@ -0,0 +1,92 @@
+{
+class CartesianImpedanceExampleController : public controller_interface::ControllerInterface
+{
+    public:
```
This formatting looks a bit off, did you run the formatter?

sp-sophia-labs replied:
done now with clang-format

AndreasKuhner commented on franka_example_controllers/include/franka_example_controllers/cartesian_impedance_example_controller.hpp:
```diff
@@ -0,0 +1,92 @@
+
+    bool k_elbow_activated{true};
+        
+    franka_msgs::msg::FrankaRobotState robot_state_, init_robot_state_;
```
One variable per line please :)

AndreasKuhner commented on franka_example_controllers/include/franka_example_controllers/cartesian_impedance_example_controller.hpp:
```diff
@@ -0,0 +1,92 @@
+    const std::string k_robot_state_interface_name{"robot_state"};
+    const std::string k_robot_model_interface_name{"robot_model"};
+
+    std::string arm_id_{"fr3"};
```
Maybe you can peek in the other example controllers on how the `arm_id` is read in the `on_init/on_configure` method (nowadays we try to make this given the actual arm_id)

AndreasKuhner commented on franka_example_controllers/include/franka_example_controllers/cartesian_impedance_example_controller.hpp:
```diff
@@ -0,0 +1,92 @@
+    Eigen::Quaterniond orientation_d_target_;   
+
+    rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr sub_equilibrium_pose_;
+    void equilibriumPoseCallback(const geometry_msgs::msg::PoseStamped::SharedPtr msg);
```
Could you please add docu to this function?

sp-sophia-labs replied:
done in .cpp just before equilibriumPoseCallback() definition

AndreasKuhner commented on franka_example_controllers/src/cartesian_impedance_example_controller.cpp:
```diff
@@ -0,0 +1,232 @@
+
+    // compute error to desired pose
+    // position error
+    Eigen::Matrix<double, 6, 1> error;
```
`error` sounds a bit generic. Maybe something like `position_error`?

AndreasKuhner commented on franka_example_controllers/include/franka_example_controllers/pseudo_inversion.hpp:
```diff
@@ -0,0 +1,21 @@
+
+namespace franka_example_controllers {
+
+inline void pseudoInverse(const Eigen::MatrixXd& M_, Eigen::MatrixXd& M_pinv_, bool damped = true) {
```
Could this be a function which just returns the inverse instead of receiving it as a parameter?

zakxbt committed:
return inverse instead of receiving parameter

zakxbt committed:
doc, variables and arm_id call fix

sp-sophia-labs commented:
@AndreasKuhner Thanks for your feedback! I have made the requested changes. 

Please let me know if there are any further adjustments needed or if there's anything else you'd like me to address

AndreasKuhner commented on franka_example_controllers/src/cartesian_impedance_example_controller.cpp:
```diff
@@ -0,0 +1,251 @@
+#include "franka_example_controllers/cartesian_impedance_example_controller.hpp"
+#include "franka_example_controllers/pseudo_inversion.hpp"
+
+const Eigen::MatrixXd IDENTITY = Eigen::MatrixXd::Identity(6, 6);
```
Are you using this somewhere? I think you always use directly the `Identity` function whereever you needed it..?

sp-sophia-labs replied:
you're right, it was an old overlap. Fixed now

zakxbt committed:
overlap clean identity

sp-sophia-labs committed:
Merge branch 'frankaemika:humble' into humble

AndreasKuhner commented:
Looks good from my side. I guess we would only need to run it on the robot for testing it :)

BarisYazici commented:
Please add the `src/cartesian_impedance_example_controller.cpp` to the `CMakeLists.txt` under `franka_example_controllers` https://github.com/frankaemika/franka_ros2/blob/d3d0fb4ca3d5beb8264ba71dab27b3c364ab1c31/franka_example_controllers/CMakeLists.txt#L29-L44

BarisYazici commented:
Add the arm_id parameter similar to the other launch file [examples](https://github.com/frankaemika/franka_ros2/blob/d3d0fb4ca3d5beb8264ba71dab27b3c364ab1c31/franka_bringup/launch/cartesian_pose_example_controller.launch.py#L33) and pass parameter this to the `franka.launch.py`

BarisYazici commented on franka_example_controllers/src/cartesian_impedance_example_controller.cpp:
```diff
@@ -0,0 +1,249 @@
+
+  // pseudoinverse for nullspace handling
+  Eigen::MatrixXd jacobian_transpose_pinv;
+  pseudoInverse(jacobian.transpose(), jacobian_transpose_pinv);
```
based on the definition on PseudoInverse from you:
```
inline Eigen::MatrixXd pseudoInverse(const Eigen::MatrixXd& M_, bool damped = true) {
  double lambda_ = damped ? 0.2 : 0.0;

  Eigen::JacobiSVD<Eigen::MatrixXd> svd(M_, Eigen::ComputeFullU | Eigen::ComputeFullV);
  Eigen::JacobiSVD<Eigen::MatrixXd>::SingularValuesType sing_vals_ = svd.singularValues();
  Eigen::MatrixXd S_ = M_;  // copying the dimensions of M_, its content is not needed.
  S_.setZero();

  for (int i = 0; i < sing_vals_.size(); i++)
    S_(i, i) = (sing_vals_(i)) / (sing_vals_(i) * sing_vals_(i) + lambda_ * lambda_);

  return svd.matrixV() * S_.transpose() * svd.matrixU().transpose();
}

```
second parameter should be `bool damped`. And the returned value should be the inverted matrix.

BarisYazici commented on franka_example_controllers/src/cartesian_impedance_example_controller.cpp:
```diff
@@ -0,0 +1,249 @@
+  tau_d << tau_task + tau_nullspace + coriolis;
+
+  // saturate the commanded torque to joint limits
+  tau_d << saturateTorqueRate(tau_d, tau_j_d);
```
typo with tau_j_d -> tau_J_d

BarisYazici commented on franka_example_controllers/src/cartesian_impedance_example_controller.cpp:
```diff
@@ -0,0 +1,249 @@
+}
+
+controller_interface::return_type CartesianImpedanceExampleController::update(
+    const rclcpp::Time& time,
```
comment time and period out to make linter happy

```suggestion

controller_interface::return_type CartesianImpedanceExampleController::update(
    const rclcpp::Time& /*time*/,
    const rclcpp::Duration& /*period*/) {
```

BarisYazici commented on franka_bringup/config/controllers.yaml:
```diff
@@ -47,6 +47,10 @@ controller_manager:
       type: franka_robot_state_broadcaster/FrankaRobotStateBroadcaster
 
+    cartesian_impedance_example_controller:
+      type: franka_example_controllers::CartesianImpedanceExampleController
```
```suggestion
      type: franka_example_controllers/CartesianImpedanceExampleController
```

BarisYazici commented on franka_example_controllers/src/cartesian_impedance_example_controller.cpp:
```diff
@@ -0,0 +1,249 @@
+
+  for (int i = 0; i < num_joints; i++) {
+    command_interfaces_[i].set_value(tau_d[i]);
+    std::cout << tau_d[i] << std::endl;
```
remove std::cout

zakxbt committed:
add executable and needed adjustement for local version

sp-sophia-labs commented:
Thanks for the review @BarisYazici. Updated

AndreasKuhner commented:
I think Baris tried to run it on the robot and it didn't work - did you do tests on the robot?

BarisYazici commented:
> I think Baris tried to run it on the robot and it didn't work - did you do tests on the robot?

Yes after fixing all the issues in the PR, it still didn't run on the robot. Did you do any tests on the robot?

sp-sophia-labs commented:
Yes, tested on FR3 and i could access the robot interface, send commands and the robot successfully converged to equilibrium pose (set it manually). I'll provide terminal outputs once i get access to the robot. 
@BarisYazici can you provide which part doesn't work, so i can help

sp-sophia-labs commented:
Here's my output :
- Successful connection with gripper
- Successful connection with FrankaHardwareInterface
- Controller loaded and configured correctly

![outputHWfr3](https://github.com/user-attachments/assets/8871e270-f9e9-4c19-b528-7525f6f92d59)

Here's a short demo : Robot starts in random pose and converges to equilibrium pose relative to starting point of fr3_link8 (published manually through /equilibrium_pose topic) 
![IMG_5002online-video-cutter com-ezgif com-optimize](https://github.com/user-attachments/assets/96f04cee-d63c-415e-bcf5-9672363ccbef)

BarisYazici commented:
> Yes, tested on FR3 and i could access the robot interface, send commands and the robot successfully converged to equilibrium pose (set it manually). I'll provide terminal outputs once i get access to the robot.
> @BarisYazici can you provide which part doesn't work, so i can help

I wrote this because the code you started the PR with was not compiling and the source code was not in the CMakelists.txt. I invested some time to fix the issues with the code and got it in a correctly compiling state. After I tested couple of time with the robot and it was throwing some communication violations. I didn't try to debug further than that. If you say you tried it on the hardware, I will give it another try. But please put all the files necessary to run the example in this PR. You mention a node to publish equilibrium pose. Could you also put it in this PR?

I noticed in the image of the terminal output your example controller name is force_impedance_control. Is it the same setup you are using in the tests? Please make sure it's exactly the same.

zakxbt committed:
add equilibrium pose publisher

AllanXuRan commented:
There is an error occurs because `franka_example_controllers` references a non-existent
  source file `equilibrium_pose_publisher.cpp` in its CMakeLists.txt. As a result,
  CMake fails with:

      Cannot find source file: src/equilibrium_pose_publisher.cpp
      No SOURCES given to target: equilibrium_pose_publisher

  How to fix this?

AndreasKuhner commented:
To my knowledge, there is no `equilibrium_pose_publisher.cpp`. So, not sure where this would come from. Maybe it was part of the old implementation?

franzesegiovanni commented:
Hello guys, I have tried to run the code from the pull request. I do not manage to have a stable behavior of the control. It always goes in safety violation for jumps in the torque or communication violation. Can anybody help in implementing a simple cartesian example controller that we can build on top? To be honest I was very happy with the ROS1 implementation that I could build on top, for the ROS2, i am struggling to have the code to even not stop due to communication constraint. Can it be that the problem is that I am running things in docker? Please help :)

souljaboy764 commented:
I had the same issue that @franzesegiovanni had wherein starting the cartesian impedance controller would also cause jumps in the torque leading to constraint violations. I added a fix for this and also integrated the Interactive Marker in rviz instead of using an equilibrium pose publisher.

TL;DR: Removing [`saturateTorqueRate`](https://github.com/sp-sophia-labs/franka_ros2/blob/64966bf59f95f9bf25446bb26ee5e2908d3c6b77/franka_example_controllers/src/cartesian_impedance_example_controller.cpp#L186) to directly use the impedance outputs (and subsequently tuning the stiffness gains).

![franka-cartesian-impedance-small-cropped](https://github.com/user-attachments/assets/77725bef-79f5-4753-8aa4-df5be7383195)

I've submitted a pull request here: sp-sophia-labs/franka_ros2#1

kmohyeldine commented:
as there is not much activity on this PR, we are taking it over and publishing it in the next release
